File: services/api-go/internal/observability/logger.go
package observability

import (
	"encoding/json"
	"fmt"
	"os"
	"time"
)

// Logger provides structured JSON logging.
type Logger struct {
	service string
}

// NewLogger creates a logger tagged with a service name.
func NewLogger(service string) *Logger {
	return &Logger{service: service}
}

func (l *Logger) log(level, msg string, fields map[string]interface{}) {
	entry := map[string]interface{}{
		"time":    time.Now().UTC().Format(time.RFC3339),
		"level":   level,
		"service": l.service,
		"msg":     msg,
	}
	for k, v := range fields {
		entry[k] = v
	}
	data, _ := json.Marshal(entry)
	fmt.Fprintln(os.Stdout, string(data))
}

// Info logs at INFO level.
func (l *Logger) Info(msg string, fields ...map[string]interface{}) {
	var f map[string]interface{}
	if len(fields) > 0 {
		f = fields[0]
	}
	l.log("info", msg, f)
}

// Warn logs at WARN level.
func (l *Logger) Warn(msg string, fields ...map[string]interface{}) {
	var f map[string]interface{}
	if len(fields) > 0 {
		f = fields[0]
	}
	l.log("warn", msg, f)
}

// Error logs at ERROR level.
func (l *Logger) Error(msg string, fields ...map[string]interface{}) {
	var f map[string]interface{}
	if len(fields) > 0 {
		f = fields[0]
	}
	l.log("error", msg, f)
}

File: services/ingest-node/src/utils/logger.js
"use strict";

/**
 * Structured JSON logger for Lightwatch services.
 * Outputs one JSON line per log entry to stdout.
 */
class Logger {
  constructor(service) {
    this.service = service;
  }

  _write(level, msg, fields = {}) {
    const entry = {
      time: new Date().toISOString(),
      level,
      service: this.service,
      msg,
      ...fields,
    };
    process.stdout.write(JSON.stringify(entry) + "\n");
  }

  info(msg, fields) {
    this._write("info", msg, fields);
  }

  warn(msg, fields) {
    this._write("warn", msg, fields);
  }

  error(msg, fields) {
    this._write("error", msg, fields);
  }

  /**
   * Returns a request logging function for the native http pipeline.
   * Call at start of request, returns a finish() function.
   */
  request(req) {
    const start = Date.now();
    const id = req.headers["x-request-id"] || randomHexId();
    return {
      id,
      finish: (statusCode) => {
        this._write("info", "request", {
          method: req.method,
          path: req.url,
          status: statusCode,
          duration_ms: Date.now() - start,
          remote: req.socket.remoteAddress,
          request_id: id,
        });
      },
    };
  }
}

function randomHexId() {
  const bytes = new Uint8Array(16);
  require("crypto").getRandomValues(bytes);
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}

module.exports = { Logger };
